usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -6750,6 +6751,135 @@ static CK_RV h_opaque_2_blob(STDLL_TokData_t *tokdata, CK_OBJECT_HANDLE handle,
+    rc = object_mgr_find_in_map1(tokdata, key, &key_obj, READ_LOCK);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Failed to acquire key from specified handle.\n");
+        if (rc == CKR_OBJECT_HANDLE_INVALID)
+            return CKR_KEY_HANDLE_INVALID;
+        else
+            return rc;
+    }
+
+    if (sign) {
+        rc = template_attribute_get_bool(key_obj->template, CKA_SIGN, &flag);
+        if (rc != CKR_OK) {
+            TRACE_ERROR("Could not find CKA_SIGN for the key.\n");
+            rc = CKR_KEY_FUNCTION_NOT_PERMITTED;
+            goto done;
+        }
+    } else {
+        rc = template_attribute_get_bool(key_obj->template, CKA_VERIFY, &flag);
+        if (rc != CKR_OK) {
+            TRACE_ERROR("Could not find CKA_VERIFY for the key.\n");
+            rc = CKR_KEY_FUNCTION_NOT_PERMITTED;
+            goto done;
+        }
+    }
+
+    if (flag != TRUE) {
+        TRACE_ERROR("%s\n", ock_err(ERR_KEY_FUNCTION_NOT_PERMITTED));
+        rc = CKR_KEY_FUNCTION_NOT_PERMITTED;
+        goto done;
+    }
+
+    if (!key_object_is_mechanism_allowed(key_obj->template, mech->mechanism)) {
+        TRACE_ERROR("Mechanism not allwed per CKA_ALLOWED_MECHANISMS.\n");
+        rc = CKR_MECHANISM_INVALID;
+        goto done;
+    }
+
+    if (mech->ulParameterLen != 0) {
+        TRACE_ERROR("%s\n", ock_err(ERR_MECHANISM_PARAM_INVALID));
+        rc = CKR_MECHANISM_PARAM_INVALID;
+        goto done;
+    }
+
+    rc = template_attribute_get_ulong(key_obj->template, CKA_KEY_TYPE,
+                                      &keytype);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Could not find CKA_KEY_TYPE for the key.\n");
+        goto done;
+    }
+
+    if (keytype != CKK_EC) {
+        TRACE_ERROR("%s\n", ock_err(ERR_KEY_TYPE_INCONSISTENT));
+        rc = CKR_KEY_TYPE_INCONSISTENT;
+        goto done;
+    }
+
+    /* Check key class: must be either a private or public key */
+    rc = template_attribute_get_ulong(key_obj->template, CKA_CLASS,
+                                      &class);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Could not find CKA_CLASS for the key.\n");
+        goto done;
+    }
+
+    if (sign) {
+        if (class != CKO_PRIVATE_KEY) {
+            TRACE_ERROR("This operation requires a private key.\n");
+            rc = CKR_KEY_FUNCTION_NOT_PERMITTED;
+            goto done;
+        }
+    } else {
+        if (class != CKO_PUBLIC_KEY) {
+            TRACE_ERROR("This operation requires a public key.\n");
+            rc = CKR_KEY_FUNCTION_NOT_PERMITTED;
+            goto done;
+        }
+    }
+
+    ctx->context_len = 0;
+    ctx->context = NULL;
+    ctx->key = key;
+    ctx->mech.ulParameterLen = mech->ulParameterLen;
+    ctx->mech.mechanism = mech->mechanism;
+    ctx->mech.pParameter = ptr;
+    ctx->multi_init = FALSE;
+    ctx->multi = FALSE;
+    ctx->active = TRUE;
+    ctx->recover = FALSE;
+    ctx->pkey_active = FALSE;
+
+    rc = CKR_OK;
+
+done:
+    object_put(tokdata, key_obj, TRUE);
```

Comment:
Why do we unlock the object here?  We just checked that everything is fine and then we allow the user to, e.g., remove CKA_SIGN and then sign with this key?  Shouldn't the unlocking be done in the _final function?

Reply:
This function obtained the lock with object_mgr_find_in_map1, so it must also release it. 
And no, OCK does not keep the lock in-between Init, Update, Final calls.